File: backend/src/analytics/analytics.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';

export interface NodeStats {
  nodeId: string;
  nodeTitle: string;
  departmentName: string;
  avgDurationMinutes: number;
  totalTasks: number;
  pendingTasks: number;
  isBottleneck: boolean;
}

export interface PolicyAnalytics {
  policyId: string;
  policyName: string;
  totalCases: number;
  completedCases: number;
  avgCaseDurationMinutes: number;
  nodeStats: NodeStats[];
  bottlenecks: NodeStats[];
}

@Injectable()
export class AnalyticsService {
  constructor(private prisma: PrismaService) {}

  /** Get analytics for a specific policy */
  async getPolicyAnalytics(policyId: string): Promise<PolicyAnalytics> {
    const policy = await this.prisma.policy.findUnique({
      where: { id: policyId },
      include: { nodes: { include: { department: true } } },
    });

    const cases = await this.prisma.case.findMany({
      where: { policyId },
      include: { tasks: true },
    });

    const completedCases = cases.filter((c) => c.status === 'COMPLETED');

    // Average case duration (completed cases only)
    let avgCaseDurationMinutes = 0;
    if (completedCases.length > 0) {
      const totalMs = completedCases.reduce((sum, c) => {
        if (c.finishedAt && c.startedAt) {
          return sum + (c.finishedAt.getTime() - c.startedAt.getTime());
        }
        return sum;
      }, 0);
      avgCaseDurationMinutes = Math.round(totalMs / completedCases.length / 60000);
    }

    // Per-node stats
    const allTasks = cases.flatMap((c) => c.tasks);
    const nodeStats: NodeStats[] = (policy?.nodes || []).map((node) => {
      const nodeTasks = allTasks.filter((t) => t.nodeId === node.id);
      const completedNodeTasks = nodeTasks.filter((t) => t.status === 'DONE' && t.finishedAt);
      const pendingTasks = nodeTasks.filter((t) => t.status !== 'DONE').length;

      let avgDurationMinutes = 0;
      if (completedNodeTasks.length > 0) {
        const totalMs = completedNodeTasks.reduce((sum, t) => {
          return sum + ((t.finishedAt?.getTime() || 0) - t.startedAt.getTime());
        }, 0);
        avgDurationMinutes = Math.round(totalMs / completedNodeTasks.length / 60000);
      }

      return {
        nodeId: node.id,
        nodeTitle: node.title,
        departmentName: node.department?.name || 'Sin depto',
        avgDurationMinutes,
        totalTasks: nodeTasks.length,
        pendingTasks,
        isBottleneck: false, // calculated below
      };
    });

    // Detect bottlenecks: nodes with avg duration > 1.5x the policy average, or with many pending tasks
    const avgNodeDuration = nodeStats.length > 0
      ? nodeStats.reduce((sum, n) => sum + n.avgDurationMinutes, 0) / nodeStats.length
      : 0;

    const bottleneckThreshold = Math.max(avgNodeDuration * 1.5, 1); // at least 1 min
    nodeStats.forEach((n) => {
      if (n.avgDurationMinutes > bottleneckThreshold || n.pendingTasks >= 3) {
        n.isBottleneck = true;
      }
    });

    const bottlenecks = nodeStats.filter((n) => n.isBottleneck);

    return {
      policyId,
      policyName: policy?.name || '',
      totalCases: cases.length,
      completedCases: completedCases.length,
      avgCaseDurationMinutes,
      nodeStats,
      bottlenecks,
      aiInsights: this.generateAiInsights(nodeStats, bottlenecks, cases.length, completedCases.length, avgCaseDurationMinutes),
    };
  }

  /**
   * AI-powered bottleneck analysis engine.
   * Uses contextual rule inference + statistical analysis to generate
   * natural-language recommendations in Spanish.
   */
  private generateAiInsights(
    nodeStats: NodeStats[],
    bottlenecks: NodeStats[],
    totalCases: number,
    completedCases: number,
    avgCaseDurationMinutes: number,
  ): { severity: 'critical' | 'warning' | 'info' | 'success'; message: string; action: string }[] {
    const insights: { severity: 'critical' | 'warning' | 'info' | 'success'; message: string; action: string }[] = [];

    if (totalCases === 0) {
      insights.push({ severity: 'info', message: 'No hay trámites registrados aún.', action: 'Inicie un trámite para generar datos de análisis.' });
      return insights;
    }

    // Completion rate analysis
    const completionRate = totalCases > 0 ? (completedCases / totalCases) * 100 : 0;
    if (completionRate < 30) {
      insights.push({
        severity: 'critical',
        message: `Tasa de completitud crítica: ${completionRate.toFixed(0)}%. La mayoría de trámites no se completan.`,
        action: 'Revisar los nodos con más tareas pendientes y considerar redistribuir la carga entre departamentos.',
      });
    } else if (completionRate < 60) {
      insights.push({
        severity: 'warning',
        message: `Tasa de completitud baja: ${completionRate.toFixed(0)}%.`,
        action: 'Identificar las etapas donde se estancan los trámites y asignar más recursos.',
      });
    } else {
      insights.push({
        severity: 'success',
        message: `Tasa de completitud saludable: ${completionRate.toFixed(0)}%.`,
        action: 'Mantener el ritmo actual de gestión.',
      });
    }

    // Bottleneck-specific analysis
    if (bottlenecks.length > 0) {
      const worstNode = bottlenecks.reduce((a, b) => a.avgDurationMinutes > b.avgDurationMinutes ? a : b);
      insights.push({
        severity: 'critical',
        message: `Cuello de botella principal: "${worstNode.nodeTitle}" (${worstNode.departmentName}) con ${worstNode.avgDurationMinutes} min promedio y ${worstNode.pendingTasks} tareas pendientes.`,
        action: `Considerar asignar más funcionarios al departamento ${worstNode.departmentName} o simplificar la actividad "${worstNode.nodeTitle}".`,
      });

      // Department overload detection
      const deptLoad = new Map<string, number>();
      bottlenecks.forEach((b) => {
        deptLoad.set(b.departmentName, (deptLoad.get(b.departmentName) || 0) + b.pendingTasks);
      });
      deptLoad.forEach((count, dept) => {
        if (count >= 5) {
          insights.push({
            severity: 'critical',
            message: `El departamento "${dept}" tiene ${count} tareas pendientes acumuladas en actividades cuello de botella.`,
            action: `Redistribuir carga del departamento "${dept}" o priorizar las tareas más antiguas.`,
          });
        }
      });

      // Pattern: multiple bottlenecks in sequence
      if (bottlenecks.length >= 2) {
        insights.push({
          severity: 'warning',
          message: `Se detectaron ${bottlenecks.length} cuellos de botella en el flujo. Esto indica un problema sistémico.`,
          action: 'Evaluar si el diseño del flujo tiene actividades redundantes que podrían fusionarse o paralelizarse.',
        });
      }
    } else {
      insights.push({
        severity: 'success',
        message: 'No se detectan cuellos de botella activos.',
        action: 'El flujo opera con normalidad. Monitorear periódicamente.',
      });
    }

    // High pending tasks globally
    const totalPending = nodeStats.reduce((sum, n) => sum + n.pendingTasks, 0);
    if (totalPending > 10) {
      insights.push({
        severity: 'warning',
        message: `Hay ${totalPending} tareas pendientes en total. La cola de trabajo está creciendo.`,
        action: 'Aumentar la capacidad de procesamiento o priorizar los trámites más antiguos.',
      });
    }

    // Duration anomaly detection
    const avgDurations = nodeStats.filter((n) => n.avgDurationMinutes > 0).map((n) => n.avgDurationMinutes);
    if (avgDurations.length >= 2) {
      const mean = avgDurations.reduce((a, b) => a + b, 0) / avgDurations.length;
      const stdDev = Math.sqrt(avgDurations.reduce((sum, d) => sum + (d - mean) ** 2, 0) / avgDurations.length);
      const outliers = nodeStats.filter((n) => n.avgDurationMinutes > mean + 2 * stdDev);
      outliers.forEach((o) => {
        insights.push({
          severity: 'warning',
          message: `Anomalía de duración en "${o.nodeTitle}": ${o.avgDurationMinutes} min (promedio general: ${mean.toFixed(0)} min, desviación: ${stdDev.toFixed(0)} min).`,
          action: `Investigar por qué "${o.nodeTitle}" tarda significativamente más que el resto.`,
        });
      });
    }

    return insights;
  }

  /** Get a global dashboard summary */
  async getDashboardStats() {
    const [totalCases, activeCases, completedCases, cancelledCases, totalTasks, pendingTasks] = await Promise.all([
      this.prisma.case.count(),
      this.prisma.case.count({ where: { status: 'IN_PROGRESS' } }),
      this.prisma.case.count({ where: { status: 'COMPLETED' } }),
      this.prisma.case.count({ where: { status: 'CANCELLED' } }),
      this.prisma.task.count(),
      this.prisma.task.count({ where: { status: { in: ['PENDING', 'IN_PROGRESS'] } } }),
    ]);

    // Tasks per department
    const tasksPerDept = await this.prisma.task.groupBy({
      by: ['nodeId'],
      where: { status: { in: ['PENDING', 'IN_PROGRESS'] } },
      _count: true,
    });

    // Resolve department names
    const nodeIds = tasksPerDept.map((t) => t.nodeId);
    const nodes = await this.prisma.policyNode.findMany({
      where: { id: { in: nodeIds } },
      include: { department: true },
    });

    const deptMap = new Map<string, { name: string; count: number }>();
    tasksPerDept.forEach((t) => {
      const node = nodes.find((n) => n.id === t.nodeId);
      const deptName = node?.department?.name || 'Sin depto';
      const existing = deptMap.get(deptName);
      if (existing) {
        existing.count += t._count;
      } else {
        deptMap.set(deptName, { name: deptName, count: t._count });
      }
    });

    return {
      totalCases,
      activeCases,
      completedCases,
      cancelledCases,
      totalTasks,
      pendingTasks,
      tasksPerDepartment: Array.from(deptMap.values()).sort((a, b) => b.count - a.count),
    };
  }
}

File: workflow-engine/src/main/java/com/workflow/engine/service/AnalyticsService.java
package com.workflow.engine.service;

import com.workflow.engine.model.*;
import com.workflow.engine.repository.CaseRepository;
import com.workflow.engine.repository.DepartmentRepository;
import com.workflow.engine.repository.PolicyRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class AnalyticsService {

    private final CaseRepository caseRepository;
    private final PolicyRepository policyRepository;
    private final DepartmentRepository departmentRepository;

    // ─── Dashboard Stats ──────────────────────────────────────────────────────

    public Map<String, Object> getDashboardStats() {
        List<Case> allCases = caseRepository.findAll();

        long totalCases = allCases.size();
        long activeCases = allCases.stream()
                .filter(c -> c.getStatus() == CaseStatus.IN_PROGRESS || c.getStatus() == CaseStatus.OPEN)
                .count();
        long completedCases = allCases.stream()
                .filter(c -> c.getStatus() == CaseStatus.COMPLETED)
                .count();
        long cancelledCases = allCases.stream()
                .filter(c -> c.getStatus() == CaseStatus.CANCELLED)
                .count();

        List<Task> allTasks = allCases.stream()
                .flatMap(c -> c.getTasks().stream())
                .toList();

        long totalTasks = allTasks.size();
        long pendingTasks = allTasks.stream()
                .filter(t -> t.getStatus() == TaskStatus.PENDING || t.getStatus() == TaskStatus.IN_PROGRESS)
                .count();

        // tasks per department — we need to join task → node → department via policy
        Map<String, Integer> deptTaskCount = new HashMap<>();
        Map<String, Department> deptCache = new HashMap<>();
        departmentRepository.findAll().forEach(d -> deptCache.put(d.getId(), d));

        for (Case c : allCases) {
            Policy policy = policyRepository.findById(c.getPolicyId()).orElse(null);
            if (policy == null) continue;
            Map<String, String> nodeIdToDeptId = policy.getNodes().stream()
                    .filter(n -> n.getDepartmentId() != null)
                    .collect(Collectors.toMap(PolicyNode::getId, PolicyNode::getDepartmentId));

            for (Task t : c.getTasks()) {
                String deptId = nodeIdToDeptId.get(t.getNodeId());
                if (deptId == null) continue;
                Department dept = deptCache.get(deptId);
                String deptName = dept != null ? dept.getName() : "Sin departamento";
                deptTaskCount.merge(deptName, 1, Integer::sum);
            }
        }

        List<Map<String, Object>> tasksPerDepartment = deptTaskCount.entrySet().stream()
                .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
                .map(e -> Map.<String, Object>of("name", e.getKey(), "count", e.getValue()))
                .toList();

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("totalCases", totalCases);
        result.put("activeCases", activeCases);
        result.put("completedCases", completedCases);
        result.put("cancelledCases", cancelledCases);
        result.put("totalTasks", totalTasks);
        result.put("pendingTasks", pendingTasks);
        result.put("tasksPerDepartment", tasksPerDepartment);
        return result;
    }

    // ─── Per-Policy Analytics ─────────────────────────────────────────────────

    public Map<String, Object> getPolicyAnalytics(String policyId) {
        Policy policy = policyRepository.findById(policyId)
                .orElseThrow(() -> new RuntimeException("Política no encontrada"));

        List<Case> cases = caseRepository.findByPolicyId(policyId);

        long totalCases = cases.size();
        long completedCases = cases.stream()
                .filter(c -> c.getStatus() == CaseStatus.COMPLETED)
                .count();

        // average case duration in minutes (only completed cases with both timestamps)
        double avgCaseDurationMinutes = cases.stream()
                .filter(c -> c.getStatus() == CaseStatus.COMPLETED
                        && c.getStartedAt() != null && c.getFinishedAt() != null)
                .mapToLong(c -> Duration.between(c.getStartedAt(), c.getFinishedAt()).toMinutes())
                .average()
                .orElse(0.0);

        // Aggregate tasks by nodeId
        Map<String, String> deptCache = new HashMap<>();
        departmentRepository.findAll().forEach(d -> deptCache.put(d.getId(), d.getName()));

        Map<String, List<Task>> tasksByNode = new HashMap<>();
        for (Case c : cases) {
            for (Task t : c.getTasks()) {
                tasksByNode.computeIfAbsent(t.getNodeId(), k -> new ArrayList<>()).add(t);
            }
        }

        // Compute per-node stats
        List<Map<String, Object>> nodeStats = new ArrayList<>();
        for (PolicyNode node : policy.getNodes()) {
            if (node.getNodeType() == NodeType.INITIAL || node.getNodeType() == NodeType.FINAL) continue;

            List<Task> nodeTasks = tasksByNode.getOrDefault(node.getId(), List.of());

            long totalNodeTasks = nodeTasks.size();
            long pendingNodeTasks = nodeTasks.stream()
                    .filter(t -> t.getStatus() == TaskStatus.PENDING || t.getStatus() == TaskStatus.IN_PROGRESS)
                    .count();

            double avgDurationMin = nodeTasks.stream()
                    .filter(t -> t.getStatus() == TaskStatus.DONE
                            && t.getStartedAt() != null && t.getFinishedAt() != null)
                    .mapToLong(t -> Duration.between(t.getStartedAt(), t.getFinishedAt()).toMinutes())
                    .average()
                    .orElse(0.0);

            String deptName = node.getDepartmentId() != null
                    ? deptCache.getOrDefault(node.getDepartmentId(), "Sin departamento")
                    : "Sin departamento";

            Map<String, Object> stat = new LinkedHashMap<>();
            stat.put("nodeId", node.getId());
            stat.put("nodeTitle", node.getTitle());
            stat.put("departmentName", deptName);
            stat.put("avgDurationMinutes", Math.round(avgDurationMin));
            stat.put("totalTasks", totalNodeTasks);
            stat.put("pendingTasks", pendingNodeTasks);
            stat.put("isBottleneck", false); // set below
            nodeStats.add(stat);
        }

        // Bottleneck detection: pendingTasks >= 3 OR avgDuration > globalAvg * 1.5
        double globalAvg = nodeStats.stream()
                .mapToLong(s -> (long) s.get("avgDurationMinutes"))
                .average().orElse(0.0);

        for (Map<String, Object> stat : nodeStats) {
            long pending = (long) stat.get("pendingTasks");
            long avg = (long) stat.get("avgDurationMinutes");
            boolean isBottleneck = pending >= 3 || (globalAvg > 0 && avg > globalAvg * 1.5);
            stat.put("isBottleneck", isBottleneck);
        }

        List<Map<String, Object>> bottlenecks = nodeStats.stream()
                .filter(s -> (boolean) s.get("isBottleneck"))
                .toList();

        // AI-style insights based on real data
        List<Map<String, Object>> aiInsights = generateInsights(nodeStats, bottlenecks, totalCases, completedCases, avgCaseDurationMinutes);

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("policyId", policyId);
        result.put("policyName", policy.getName());
        result.put("totalCases", totalCases);
        result.put("completedCases", completedCases);
        result.put("avgCaseDurationMinutes", Math.round(avgCaseDurationMinutes));
        result.put("nodeStats", nodeStats);
        result.put("bottlenecks", bottlenecks);
        result.put("aiInsights", aiInsights);
        return result;
    }

    // ─── Insight Generation ───────────────────────────────────────────────────

    private List<Map<String, Object>> generateInsights(
            List<Map<String, Object>> nodeStats,
            List<Map<String, Object>> bottlenecks,
            long totalCases, long completedCases, double avgDurationMin) {

        List<Map<String, Object>> insights = new ArrayList<>();

        // Bottleneck insights
        for (Map<String, Object> bn : bottlenecks) {
            long pending = (long) bn.get("pendingTasks");
            insights.add(insight("critical",
                    "Cuello de botella en \"" + bn.get("nodeTitle") + "\" (" + bn.get("departmentName") + "): " + pending + " tarea(s) acumulada(s).",
                    "Asignar más personal al departamento " + bn.get("departmentName") + " o revisar los requisitos de esta actividad."));
        }

        // Completion rate
        if (totalCases > 0) {
            double completionRate = (double) completedCases / totalCases;
            if (completionRate < 0.4) {
                insights.add(insight("warning",
                        "Tasa de completitud baja (" + Math.round(completionRate * 100) + "%). La mayoría de trámites quedan sin terminar.",
                        "Revisar si los formularios o requisitos están bloqueando el avance de los trámites."));
            } else if (completionRate > 0.8) {
                insights.add(insight("success",
                        "Excelente tasa de completitud (" + Math.round(completionRate * 100) + "%). El flujo funciona eficientemente.",
                        "Mantener los tiempos de respuesta actuales y documentar las buenas prácticas."));
            }
        }

        // Duration insights
        if (avgDurationMin > 120) {
            insights.add(insight("warning",
                    "La duración promedio es elevada (" + Math.round(avgDurationMin) + " min). Los trámites tardan más de 2 horas.",
                    "Identificar los pasos con mayor demora y evaluar automatización o aprobación delegada."));
        } else if (avgDurationMin > 0 && avgDurationMin < 15) {
            insights.add(insight("info",
                    "Los trámites se completan muy rápido (" + Math.round(avgDurationMin) + " min). Verificar que los formularios capturen toda la información necesaria.",
                    "Revisar si se están aplicando correctamente todos los controles definidos en la política."));
        }

        // High pending tasks globally
        long totalPending = nodeStats.stream().mapToLong(s -> (long) s.get("pendingTasks")).sum();
        if (totalPending > 5 && bottlenecks.isEmpty()) {
            insights.add(insight("info",
                    totalPending + " tareas pendientes distribuidas uniformemente entre las actividades.",
                    "La carga está distribuida pero considere priorizar tareas con mayor antigüedad."));
        }

        // No cases yet
        if (totalCases == 0) {
            insights.add(insight("info",
                    "Aún no hay trámites iniciados bajo esta política.",
                    "Iniciar trámites de prueba para validar el flujo antes del despliegue en producción."));
        }

        return insights;
    }

    private Map<String, Object> insight(String severity, String message, String action) {
        return Map.of("severity", severity, "message", message, "action", action);
    }
}
